multipartfile.go

```diff
@@ -22,94 +21,111 @@ const (
 	contentTypeHeader = "Content-Type"
 )
 
-var ErrPartOutsideParent = errors.New("file outside parent dir")
-var ErrPartInChildTree = errors.New("file in child tree")
+type multipartDirectory struct {
+	path   string
+	walker *multipartWalker
 
-// multipartFile implements Node, and is created from a `multipart.Part`.
-type multipartFile struct {
-	Node
+	// part is the part describing the directory. It's nil when implicit.
+	part *multipart.Part
+}
 
-	part      *multipart.Part
-	reader    *peekReader
-	mediatype string
+type multipartWalker struct {
+	part   *multipart.Part
+	reader *multipart.Reader
 }
 
-func NewFileFromPartReader(reader *multipart.Reader, mediatype string) (Directory, error) {
-	if !isDirectory(mediatype) {
-		return nil, ErrNotDirectory
-	}
+func (m *multipartWalker) consumePart() {
+	m.part = nil
+}
 
-	f := &multipartFile{
-		reader:    &peekReader{r: reader},
-		mediatype: mediatype,
+func (m *multipartWalker) getPart() (*multipart.Part, error) {
+	if m.part != nil {
+		return m.part, nil
+	}
+	if m.reader == nil {
+		return nil, io.EOF
 	}
 
-	return f, nil
+	var err error
+	m.part, err = m.reader.NextPart()
+	if err == io.EOF {
+		m.reader = nil
+	}
+	return m.part, err
 }
 
-func newFileFromPart(parent string, part *multipart.Part, reader *peekReader) (string, Node, error) {
-	f := &multipartFile{
-		part:   part,
-		reader: reader,
+func NewFileFromPartReader(reader *multipart.Reader, mediatype string) (Directory, error) {
+	if !isDirectory(mediatype) {
+		return nil, ErrNotDirectory
 	}
 
-	dir, base := path.Split(f.fileName())
-	dir = path.Clean(dir)
-	parent = path.Clean(parent)
-	if dir == "." {
-		dir = ""
-	}
-	if parent == "." {
-		parent = ""
-	}
+	return &multipartDirectory{
+		path: "/",
+		walker: &multipartWalker{
+			reader: reader,
+		},
+	}, nil
+}
 
-	if dir != parent {
-		if strings.HasPrefix(dir, parent) {
-			return "", nil, ErrPartInChildTree
-		}
-		return "", nil, ErrPartOutsideParent
+func nextFile(w *multipartWalker) (Node, error) {
```

Comment:
I moved everything related to paths _out_ of this function. Otherwise, we would have had to do a bunch of path parsing twice.